File: peacedesign/src/main/java/com/peacedesign/android/utils/touchutils/HoverPushEffect.java
package com.peacedesign.android.utils.touchutils;

import android.animation.ValueAnimator;
import android.view.MotionEvent;
import android.view.View;
import android.view.View.OnTouchListener;

import androidx.annotation.NonNull;

public class HoverPushEffect implements OnTouchListener {
    private final float pressure;
    private float downScale;

    public HoverPushEffect() {
        this(Pressure.MEDIUM);
    }

    public HoverPushEffect(@NonNull Pressure pressure) {
        this.pressure = pressure == Pressure.LOW ? 0.98f : pressure == Pressure.HIGH ? 0.90f : 0.95f;
    }

    @Override
    public boolean onTouch(@NonNull View v, @NonNull MotionEvent e) {
        boolean defaultResult = v.onTouchEvent(e);
        float x;
        float y;
        switch (e.getAction()) {
            case MotionEvent.ACTION_DOWN:
                down(v);
                break;
            case MotionEvent.ACTION_UP:
                up(v);
                x = e.getX();
                y = e.getY();
                if (!defaultResult && (x >= 0 && y >= 0 && x <= v.getWidth() && y <= v.getHeight())) v.performClick();
                break;
            case MotionEvent.ACTION_CANCEL:
                up(v); break;
            default:
                return defaultResult;
        }
        return true;
    }

    protected void down(@NonNull View v) {
        downScale = v.getScaleX() * pressure;
        animate(v, 1f, downScale);
    }

    protected void up(@NonNull View v) {
        animate(v, downScale, 1f);
    }

    private void animate(View v, float from, float to) {
        ValueAnimator animator = ValueAnimator.ofFloat(from, to);
        animator.setDuration(100);
        animator.addUpdateListener(animation -> {
            float value = (float) animation.getAnimatedValue();
            v.setScaleX(value);
            v.setScaleY(value);
        });
        animator.start();
    }

    public void onLongPress() { }

    public enum Pressure {
        LOW, MEDIUM, HIGH,
    }
}

File: app/src/main/java/com/quranapp/android/utils/gesture/HoverPushEffect.kt
package com.quranapp.android.utils.gesture

import android.animation.ValueAnimator
import android.view.MotionEvent
import android.view.View

open class HoverPushEffect(pressure: Pressure = Pressure.MEDIUM) : View.OnTouchListener {
    private var hoverPressure = 0f
    private var downScale = 0f

    init {
        when (pressure) {
            Pressure.LOW -> this.hoverPressure = 0.98f
            Pressure.MEDIUM -> this.hoverPressure = 0.95f
            Pressure.HIGH -> this.hoverPressure = 0.90f
        }
    }

    override fun onTouch(v: View, e: MotionEvent): Boolean {
        val superResult = v.onTouchEvent(e)
        when (e.action) {
            MotionEvent.ACTION_DOWN -> down(v)
            MotionEvent.ACTION_UP -> {
                up(v)
                val x = e.x
                val y = e.y
                if (!superResult && (x >= 0 && y >= 0 && x <= v.width && y <= v.height)) v.performClick()
            }
            MotionEvent.ACTION_CANCEL -> up(v)
            else -> return superResult
        }
        return true
    }

    protected open fun down(v: View) {
        downScale = v.scaleX * hoverPressure
        animate(v, 1f, downScale)
    }

    protected open fun up(v: View) {
        animate(v, downScale, 1f)
    }

    private fun animate(v: View, from: Float, to: Float) {
        ValueAnimator.ofFloat(from, to).apply {
            duration = 100
            addUpdateListener { animation: ValueAnimator ->
                val value = animation.animatedValue as Float
                v.scaleX = value
                v.scaleY = value
            }
        }.start()
    }

    enum class Pressure {
        LOW, MEDIUM, HIGH
    }
}
